[COLAB-2231] Promote to Owner

umov opened the pull request:
This PR implements the "Promote to Owner" feature on the proposal's team tab.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cci-mit/xcolab/30)
<!-- Reviewable:end -->

jobachh commented:
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[microservices/clients/contestproposal-client/src/main/java/org/xcolab/client/proposals/ProposalClient.java, line 207 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-KtxkTp_ugk78QYil48i:-KtxkTp_ugk78QYil48j:b-xk8sc9) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/microservices/clients/contestproposal-client/src/main/java/org/xcolab/client/proposals/ProposalClient.java#L207)):*
> ```Java
>     }
> 
>     public void promoteUserToProposalOwner(Long proposalId, Long memberUserId) {
> ```

Let's use the member terminology: `promoteMemberToProposalOwner` and `memberId`

---

*[microservices/clients/contestproposal-client/src/main/java/org/xcolab/client/proposals/ProposalClient.java, line 328 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-Ktxkclei-kmMcqrg8xH:-Ktxkclfd1eAsIzVo6wN:b-axe9yn) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/microservices/clients/contestproposal-client/src/main/java/org/xcolab/client/proposals/ProposalClient.java#L328)):*
> ```Java
> 
> 
>     public List<Proposal> getProposalLinks(Long proposalId) {
> ```

This looks like an unrelated change

---

*[microservices/services/contest-service/src/main/java/org/xcolab/service/proposal/service/proposal/ProposalService.java, line 332 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-KtxmYPCdLqLAJKAbZiW:-KtxmYPCdLqLAJKAbZiX:bko832y) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/microservices/services/contest-service/src/main/java/org/xcolab/service/proposal/service/proposal/ProposalService.java#L332)):*
> ```Java
>             throw new ProposalNotFoundException("Proposal with id : " + proposalId + " not found.");
>         }
>     }
> ```

Good use of the service :+1: 

---

*[microservices/services/contest-service/src/main/java/org/xcolab/service/proposal/web/ProposalsController.java, line 229 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-KtxmOFgCQmX6sVwviwl:-KtxmOFgCQmX6sVwviwm:b-py7m29) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/microservices/services/contest-service/src/main/java/org/xcolab/service/proposal/web/ProposalsController.java#L229)):*
> ```Java
>     }
> 
>     @RequestMapping(value = "/proposals/{proposalId}/promoteUserToProposalOwner", method = RequestMethod.POST)
> ```

You can use `@PostMapping` here

---

*[microservices/services/contest-service/src/main/java/org/xcolab/service/proposal/web/ProposalsController.java, line 238 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-Ktxm2hbsFG5DHXhtjh-:-Ktxm2hbsFG5DHXhtjh0:bavggbb) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/microservices/services/contest-service/src/main/java/org/xcolab/service/proposal/web/ProposalsController.java#L238)):*
> ```Java
>         } catch (ProposalNotFoundException ignored) {
>         }
>         throw new NotFoundException();
> ```

In this case, it's probably clearer to have the throw statement in the catch clause as that's the only time it will be executed (and that way the catch class doesn't need to be `ignored`).

---

*[view/src/main/java/org/xcolab/view/pages/proposals/view/proposal/tabs/ProposalTeamTabController.java, line 47 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-Ktxku6dFObp0-JBQjjR:-Ktxku6dFObp0-JBQjjS:bvrc04i) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/view/src/main/java/org/xcolab/view/pages/proposals/view/proposal/tabs/ProposalTeamTabController.java#L47)):*
> ```Java
>         model.addAttribute("requestMembershipInviteBean", new RequestMembershipInviteBean());
> 
>         List<Proposal> listOfLinkedProposals = proposalClient.getProposalLinks(proposalId);
> ```

This seems unrelated, but I  do agree that the naming is confusing since "linking" proposals are assigned to a variable for "linked" proposals. The name change to "Links" might not help clear up this confusion though, as there is the concept of both "linking" and "linked", so it's unclear which one "links" is referring to. Let's separate this out of this PR though.

---

*[view/src/main/java/org/xcolab/view/pages/proposals/view/proposal/tabs/ProposalTeamTabController.java, line 65 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-KtxnRgYpF2TaBXolem5:-KtxnRgYpF2TaBXolem6:b-rtbq50) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/view/src/main/java/org/xcolab/view/pages/proposals/view/proposal/tabs/ProposalTeamTabController.java#L65)):*
> ```Java
>             throws ProposalsAuthorizationException, IOException {
>         checkHasManagePermissions(proposalContext, actingMember);
>         checkIsOwnerRemoved(proposalContext.getProposal(), memberId);
> ```

Great refactoring here - this is much clearer :+1: 

---

*[view/src/main/java/org/xcolab/view/pages/proposals/view/proposal/tabs/ProposalTeamTabController.java, line 120 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-KtxmxBfe6nymB9bI1hU:-KtxmxBfe6nymB9bI1hV:b-r3k1yg) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/view/src/main/java/org/xcolab/view/pages/proposals/view/proposal/tabs/ProposalTeamTabController.java#L120)):*
> ```Java
>     }
> 
>     private void checkIsOwnerRemoved(Proposal proposal, long removedMemberId)
> ```

It sounds like it checks if the proposal owner is already removed. A better name might be `checkIsRemovingOwner`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cci-mit/xcolab/30)*
<!-- Sent from Reviewable.io -->

umov commented:
Review status: 2 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[microservices/clients/contestproposal-client/src/main/java/org/xcolab/client/proposals/ProposalClient.java, line 207 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-KtxkTp_ugk78QYil48i:-Ku0RAaVOjbvo7RNATUf:b-sq86fx) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/microservices/clients/contestproposal-client/src/main/java/org/xcolab/client/proposals/ProposalClient.java#L207)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

Let's use the member terminology: `promoteMemberToProposalOwner` and `memberId`
</blockquote></details>

Done.
I've replaced `user` with `member` in all methods related to `promoteToProposalOwner` and `removeFromProposalTeam`, because this PR touches both, and we don't want inconsistency.

---

*[microservices/services/contest-service/src/main/java/org/xcolab/service/proposal/web/ProposalsController.java, line 229 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-KtxmOFgCQmX6sVwviwl:-Ku0jE0KhA6zTVUOUZDA:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/microservices/services/contest-service/src/main/java/org/xcolab/service/proposal/web/ProposalsController.java#L229)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

You can use `@PostMapping` here
</blockquote></details>

Done.

---

*[microservices/services/contest-service/src/main/java/org/xcolab/service/proposal/web/ProposalsController.java, line 238 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-Ktxm2hbsFG5DHXhtjh-:-Ku0jR9QnzTH1xKrb928:blsxqxt) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/microservices/services/contest-service/src/main/java/org/xcolab/service/proposal/web/ProposalsController.java#L238)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

In this case, it's probably clearer to have the throw statement in the catch clause as that's the only time it will be executed (and that way the catch class doesn't need to be `ignored`).
</blockquote></details>

Done. Applied the same change to `removeMemberFromProposalTeam`.

---

*[view/src/main/java/org/xcolab/view/pages/proposals/view/proposal/tabs/ProposalTeamTabController.java, line 47 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-Ktxku6dFObp0-JBQjjR:-Ku0jgo3Sn8ToCakusLH:boi8sk4) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/view/src/main/java/org/xcolab/view/pages/proposals/view/proposal/tabs/ProposalTeamTabController.java#L47)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

This seems unrelated, but I  do agree that the naming is confusing since "linking" proposals are assigned to a variable for "linked" proposals. The name change to "Links" might not help clear up this confusion though, as there is the concept of both "linking" and "linked", so it's unclear which one "links" is referring to. Let's separate this out of this PR though.
</blockquote></details>

It actually was kind of related. I was confused by the name `getLinkingProposalsForProposalID`, so I wanted to change it to `getLinkingProposals` (because this is not ObjC). However, that signature is already taken. What is the difference between `getLinkingProposalsForProposalID` and `getLinkingProposals`? For now, I've removed the rename from this PR.

---

*[view/src/main/java/org/xcolab/view/pages/proposals/view/proposal/tabs/ProposalTeamTabController.java, line 120 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-KtxmxBfe6nymB9bI1hU:-Ku0k6EXGdKX_rZ4PAYm:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/view/src/main/java/org/xcolab/view/pages/proposals/view/proposal/tabs/ProposalTeamTabController.java#L120)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

It sounds like it checks if the proposal owner is already removed. A better name might be `checkIsRemovingOwner`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cci-mit/xcolab/30)*
<!-- Sent from Reviewable.io -->

jobachh commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[view/src/main/java/org/xcolab/view/pages/proposals/view/proposal/tabs/ProposalTeamTabController.java, line 47 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/30#-Ktxku6dFObp0-JBQjjR:-Ku1pDTwA5teWfQA16vs:bmf13xt) ([raw file](https://github.com/cci-mit/xcolab/blob/9b8054dffbaea3ff3240515c9f30f122e91705b5/view/src/main/java/org/xcolab/view/pages/proposals/view/proposal/tabs/ProposalTeamTabController.java#L47)):*
<details><summary><i>Previously, umov wrote…</i></summary><blockquote>

It actually was kind of related. I was confused by the name `getLinkingProposalsForProposalID`, so I wanted to change it to `getLinkingProposals` (because this is not ObjC). However, that signature is already taken. What is the difference between `getLinkingProposalsForProposalID` and `getLinkingProposals`? For now, I've removed the rename from this PR.
</blockquote></details>

Good call about these confusing names - we'll need to clean that up. Can you create a ticket to improve the linked/linking proposals code?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cci-mit/xcolab/30)*
<!-- Sent from Reviewable.io -->
